From: Mel Gorman <mgorman@suse.de>
Date: Thu, 26 May 2011 14:53:52 +0100
Subject: [PATCH] mm: thp: Do not take expensive steps in direct reclaim for transparent hugepages
References: Reclaim/compaction (fate#311931)
Patch-mainline: no

If memory is mostly used by anonymous pages, THP can cause pages to be
unmapped, excessive memory to reclaim and processes to stall. This is
undesirable. Avoid expensive steps being taken for THP in low memory
situations.

Signed-off-by: Mel Gorman <mgorman@suse.de>
---
 mm/vmscan.c |    7 ++++---
 1 files changed, 4 insertions(+), 3 deletions(-)

diff --git a/mm/vmscan.c b/mm/vmscan.c
index 06d91a2..fab5266 100644
--- a/mm/vmscan.c
+++ b/mm/vmscan.c
@@ -2014,7 +2014,9 @@ static unsigned long do_try_to_free_pages(struct zonelist *zonelist,
 			}
 		}
 		total_scanned += sc->nr_scanned;
-		if (sc->nr_reclaimed >= sc->nr_to_reclaim)
+		if (sc->nr_reclaimed >= sc->nr_to_reclaim ||
+				((sc->gfp_mask & __GFP_NO_KSWAPD) &&
+					(total_scanned > (4UL << sc->order) || !sc->nr_reclaimed)))
 			goto out;
 
 		/*
@@ -2066,8 +2067,8 @@ unsigned long try_to_free_pages(struct zonelist *zonelist, int order,
 		.gfp_mask = gfp_mask,
 		.may_writepage = !laptop_mode,
 		.nr_to_reclaim = SWAP_CLUSTER_MAX,
-		.may_unmap = 1,
-		.may_swap = 1,
+		.may_unmap = !(gfp_mask & __GFP_NO_KSWAPD),
+		.may_swap = !(gfp_mask & __GFP_NO_KSWAPD),
 		.swappiness = vm_swappiness,
 		.order = order,
 		.mem_cgroup = NULL,
